x-pack/plugins/lens/public/editor_frame_service/editor_frame/workspace_panel/workspace_panel.tsx

```diff
@@ -232,31 +232,47 @@ export const InnerWorkspacePanel = React.memo(function InnerWorkspacePanel({
       if (renderDeps.current) {
         const [defaultLayerId] = Object.keys(renderDeps.current.datasourceLayers);
 
-        const requestWarnings: string[] = [];
-        const datasource = Object.values(renderDeps.current?.datasourceMap)[0];
-        const datasourceState = Object.values(renderDeps.current?.datasourceStates)[0].state;
-        if (adapters?.requests) {
-          plugins.data.search.showWarnings(adapters.requests, (warning) => {
-            const warningMessage = datasource.getSearchWarningMessages?.(datasourceState, warning);
+        const warningsMap: Map<string, Array<string | React.ReactNode>> = new Map();
+        const datasource = Object.values(renderDeps.current.datasourceMap)[0];
+        const datasourceState = Object.values(renderDeps.current.datasourceStates)[0].state;
 
-            requestWarnings.push(...(warningMessage || []));
-            if (warningMessage && warningMessage.length) return true;
+        if (adapters?.requests) {
+          plugins.data.search.showWarnings(adapters.requests, (warning, meta) => {
+            const { request, response, requestId } = meta;
+
+            const warningMessages = datasource.getSearchWarningMessages?.(
+              datasourceState,
+              warning,
+              request,
+              response
+            );
+
+            if (warningMessages?.length) {
+              const key = (requestId ?? '') + warning.type + warning.reason?.type ?? '';
+              if (!warningsMap.has(key)) {
+                warningsMap.set(key, warningMessages);
+              }
+              return true;
+            }
+            return false;
           });
         }
-        if (adapters && adapters.tables) {
-          dispatchLens(
-            onActiveDataChange({
-              activeData: Object.entries(adapters.tables?.tables).reduce<Record<string, Datatable>>(
-                (acc, [key, value], index, tables) => ({
-                  ...acc,
-                  [tables.length === 1 ? defaultLayerId : key]: value,
-                }),
-                {}
-              ),
-              requestWarnings,
-            })
-          );
-        }
+
+        dispatchLens(
+          onActiveDataChange({
```

Comment:
Could you explain why we need to dispatch `onActiveDataChange` regardless whether we actually have active data? (I'm sure there was a reason, but I couldn't see it just looking at the code.)

Reply:
We need to have an up to date list of `warningMessages` in the state, so some action should be called at the end.  

It seems to me that `onActiveDataChange` is suitable for this. Please see also changes in `x-pack/plugins/lens/public/state_management/lens_slice.ts`. Now activeData and requestWarnings are optional